index.ts

```diff
@@ -41,22 +45,38 @@ const CUSTOMERS = [
   },
 ];
 
+
+
 // create a function that receives a name of a customer, and returns the balance of that customer.
 // if type of this function is not infered, type it.
-
 // EXTRA: optimize this fucnction to decrease the time complexity.
 // EXTRA 2 : optimize this function to be able to memorize the results of previous calls.
-
 // i: getBalance('John')
 // o: 125
 
-// consider the following code:
 
-const z = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
-const r = Math.ceil(Math.random() * 10);
+const obtener_saldo = (() => {
+const cache = new Map<string, number>();
+     return (nombre: string): string => {
+      if (cache.has(nombre)) {
+      return cache.get(nombre)!.toString();
+      }
+      const cliente = CUSTOMERS.find((c) => c.customer === nombre);
+       if (cliente) {
+       cache.set(nombre, cliente.balance);
+       return cliente.balance.toString();
+       } else {
+       return "cliente no encontrado";
+    }
+  };
+})();
 
+console.log(obtener_saldo("John")); 
+
+
+var z: number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+var r: number = Math.floor(Math.random() * z.length);
```

Comment:
restore the original code in this section

Reply:
se ha regresado al estado original